api/layer/object.go

```diff
@@ -478,20 +484,35 @@ func (n *layer) listSortedObjectsFromNeoFS(ctx context.Context, p allObjectParam
 	return objects, nil
 }
 
-func (n *layer) getAllObjectsVersions(ctx context.Context, bkt *cache.BucketInfo, prefix, delimiter string) (map[string]*objectVersions, error) {
-	ids, err := n.objectSearch(ctx, &findParams{cid: bkt.CID})
-	if err != nil {
-		return nil, err
+func (n *layer) getAllObjectsVersions(ctx context.Context, bkt *api.BucketInfo, prefix, delimiter string) (map[string]*objectVersions, error) {
+	var (
+		err            error
+		idsListChanged = false
+	)
+
+	cacheKey := cache.CreateObjectsListCacheKey(bkt.CID, prefix)
+	ids := n.listsCache.Get(cacheKey)
+
+	if ids == nil {
+		ids, err = n.objectSearch(ctx, &findParams{cid: bkt.CID, prefix: prefix})
+		if err != nil {
+			return nil, err
+		}
+		n.listsCache.Put(cacheKey, ids)
 	}
 
 	versions := make(map[string]*objectVersions, len(ids)/2)
-	for _, id := range ids {
-		meta, err := n.objectHead(ctx, bkt.CID, id)
-		if err != nil {
-			n.log.Warn("could not fetch object meta", zap.Error(err))
+
+	for i := 0; i < len(ids); i++ {
+		obj := n.objectFromObjectsCacheOrNeoFS(ctx, bkt.CID, ids[i])
+		if obj == nil {
+			// remove an invalid object's address from list in cache
+			ids = append(ids[:i], ids[i+1:]...)
```

Comment:
Maybe we can do something like:
```
validIds := make([]*object.ID, 0, len(ids)
for _, id  := range ids {
	obj := n.objectFromObjectsCacheOrNeoFS(ctx, bkt.CID, ids[i])
	if obj == nil {
		continue
	}
        validIds := append(validIds, id)
  //...
}

if len(validIds) != len(ids) {
    // ...
}
```
or we want to save memory?

Reply:
Or we can do nothing because

Reply:
Changed implementation, now it marks  addresses to obviously invalid objects.